pkg/operator/ceph/object/dependents.go

```diff
@@ -0,0 +1,115 @@
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	v1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/util/dependents"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const bucketDependentType = "buckets in the object store (could be from ObjectBucketClaims or COSI Buckets)"
+
+// CephObjectStoreDependents returns the buckets which exist in the object store that should block
+// deletion.
+// TODO: need unit tests for this - need to be able to fake the admin ops API (nontrivial)
+func CephObjectStoreDependents(
+	clusterdCtx *clusterd.Context,
+	clusterInfo *client.ClusterInfo,
+	store *v1.CephObjectStore,
+) (*dependents.DependentList, error) {
+	nsName := fmt.Sprintf("%s/%s", store.Namespace, store.Name)
+	baseErrMsg := fmt.Sprintf("failed to get dependents of CephObjectStore %q", nsName)
+
+	deps := dependents.NewDependentList()
+
+	// NOTE: we should still check for buckets when the RGW connection is external since we have no
+	// way of knowing if the bucket was created due to an ObjectBucketClaim or COSI Bucket.
+	err := getBucketDependents(deps, clusterdCtx, clusterInfo, store)
+	if err != nil {
+		return deps, errors.Wrapf(err, baseErrMsg)
+	}
+
+	// CephObjectStoreUsers
+	users, err := clusterdCtx.RookClientset.CephV1().CephObjectStoreUsers(store.Namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return deps, errors.Wrapf(err, "%s. failed to list CephObjectStoreUsers for CephObjectStore %q", baseErrMsg, nsName)
+	}
+	for _, user := range users.Items {
+		if user.Spec.Store == store.Name {
+			deps.Add("CephObjectStoreUsers", user.Name)
+		}
+		logger.Debugf("found CephObjectStoreUser %q that does not depend on CephObjectStore %q", user.Name, nsName)
+	}
+
+	return deps, nil
+}
+
+// adds bucket dependents to the given dependents list
+func getBucketDependents(
+	deps *dependents.DependentList,
+	clusterdCtx *clusterd.Context,
+	clusterInfo *client.ClusterInfo,
+	store *v1.CephObjectStore,
+) error {
+	nsName := fmt.Sprintf("%s/%s", store.Namespace, store.Name)
+
+	objCtx, err := NewMultisiteContext(clusterdCtx, clusterInfo, store)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get object context")
+	}
+
+	missingPools, err := missingPools(objCtx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list pools in the object store")
+	}
+	if len(missingPools) > 0 {
+		// this may be an external object store that does not have all the necessary pools.
+		// this may also be a Rook-created store that did not finish deleting all pools before the
+		// Rook operator restarted.
+		// in either case, we cannot get a successful connection to RGW(s) to check for buckets, and
+		// we can assume it is safe for deletion to proceed
+		logger.Infof("skipping check for bucket dependents of CephObjectStore %q. some pools are missing: %v", nsName, missingPools)
+		return nil
```

Comment:
I have one another wrt external cluster, If the user doesn't want to remove buckets in the external cluster, but want to delete existing `cephobjectstore` CR. This PR will block the deletion of `CephObjectStore` CR right? I am not sure what is the right approach/policy for the external cluster(we can argue OBC/CephObjectStores CRDs need to removed before the cephobjectstore deletion but IMO other backend buckets should be unaffected). Any thoughts @BlaineEXE @leseb @travisn ??

Reply:
It's an interesting question, doesn't bucket removal depend on the deletion policy associated with the StorageClass?